lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,175 @@
+    end
+
+    def self.all
+      child_classes.dup
+    end
+
+    def self.discover(context:, logger:)
+      logger.info("Discovering custom resources:")
+      @resources = nil
+      @child_classes = nil
+      discover_tpr(v1beta1_kubeclient(context))
+      begin
+        discover_crd(v1beta1_crd_kubeclient(context))
+      rescue KubeException => err
+        logger.warn("Unable to discover CustomResourceDefinitions: #{err}")
+      end
+    end
+
+    def self.build(namespace:, context:, definition:, logger:)
+      # We only discover once per kubernetes-deploy invocation
+      discover(context: context, logger: logger) unless @resources
+
+      type = definition["kind"]
+      group, _, version = definition['apiVersion'].rpartition('/')
+
+      resource_class = @resources.dig(group, type, version) if @resources
+      return super unless resource_class
+
+      opts = { namespace: namespace, context: context, definition: definition, logger: logger }
+      resource_class.new(**opts)
+    end
+
+    def self.discover_tpr(client)
+      return unless client.respond_to? :get_third_party_resources
+      resources = client.get_third_party_resources
+      resources.each do |res|
+        type, _, group = res.metadata.name.partition('.')
+        # TPR API supports multiple versions for a single resource :(
+        res.versions.each do |version|
+          discovered(group: group,
+                     type: type,
+                     version: version.name,
+                     annotations: res.metadata.annotations)
+        end
+      end
+    end
+
+    def self.discover_crd(client)
+      return unless client.respond_to? :get_custom_resource_definitions
+      resources = client.get_custom_resource_definitions
+      resources.each do |res|
+        discovered(group: res.spec.group,
+                   type: res.spec.names.kind,
+                   version: res.spec.version,
+                   annotations: res.metadata.annotations)
+      end
+    end
+
+    def self.discovered(group:, type:, version:, annotations:)
+      resource_class = Class.new(self) do
+        type = type.capitalize # TPRs are inconsistent about capitalization
+        @group = group
+        @type = type
+        @version = version
+        @prunable = DiscoverableResource.parse_bool(annotations[PRUNABLE_ANNOTATION])
+        @predeploy = DiscoverableResource.parse_bool(annotations[PREDEPLOY_ANNOTATION])
+        @timeout = DiscoverableResource.parse_timeout(type, annotations[TIMEOUT_ANNOTATION])
+
+        status_field = annotations[STATUS_FIELD_ANNOTATION]
+        success_status = annotations[STATUS_SUCCESS_ANNOTATION]
+
+        define_method 'deploy_succeeded?' do
+          getter = "get_#{type.downcase}"
+          @client ||= DiscoverableResource.kubeclient(context: @context, resource_class: self.class)
+          raw_json = @client.send(getter, @name, @namespace, as: :raw)
+          query_path = JsonPath.new(status_field)
```

Comment:
We allow the CRD implementors to specify how the resources should be queried via arbitrary `jsonpath` to determine its status.